### PR TITLE
add error codes for overloaded and internal error to participant state

### PR DESCRIPTION
**Pull request by mziolekda**
The overloaded and internal error codes currently missing from the participant state API. Unfortunately, these two are quite common errors for gRPC based ledger implementations, where back pressure can result in overloaded and any 3rd party library crash will result in internal error being raised.

These error codes will very soon be replaced by SubmissionResult in a separate PR, which already contains the correct codes.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by SamirTalwar** [comment]
Worth adding a test to the _ledger-api-test-tool_?

**Comment by mziolekda** [comment]
@SamirTalwar It did cross my mind, however, it is almost impossible to find a common denominator i.e. a test case that will invariably fail with a certain stimulus
- Overloaded: different implementations of the ledgers will react differently to request overload. Those based on remote ledgers will have different overload limits depending on underlying technology, queues and settings. InMemory implementation will never respond with an overload
- InternalError: even harder task here. Internal error happens because implementation veered off the charted territory covered by proper semantic errors. Usually, this means a stack dump. I would hope we cannot construct a test case that invariably forces all integrations to crash

**Comment by SamirTalwar** [comment]
Makes sense to me.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PartyAllocationResult.scala [hunk 1]
```diff
@@ -16,19 +16,30 @@ object PartyAllocationResult {
...
+
   /** Synchronous party allocation is not supported */
   final case object NotSupported extends PartyAllocationResult {
     override def description: String = "Party allocation not supported"
   }
 
+  /** Submission ended up with internal error */
+  final case class InternalError(reason: String) extends PartyAllocationResult {
+    override def description: String =
+      "Party allocation failed with an internal error, reason=" + reason
```
Comment (miklos-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PartyAllocationResult.scala]: You may want to use string interpolation instead.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PartyAllocationResult.scala [hunk 2]
```diff
@@ -16,19 +16,30 @@ object PartyAllocationResult {
     override def description: String = "Party successfully allocated"
   }
 
+  /** The system is overloaded, clients should back off exponentially */
+  final case object Overloaded extends PartyAllocationResult {
+    override def description: String = "System is overloaded, please try again later"
```
Comment (stefanobaghino-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PartyAllocationResult.scala]:
```suggestion
    override val description: String = "System is overloaded, please try again later"
```
Not sure it makes sense to wrap a constant around a method.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PartyAllocationResult.scala [hunk 3]
```diff
@@ -16,19 +16,30 @@ object PartyAllocationResult {
     override def description: String = "Party successfully allocated"
   }
 
+  /** The system is overloaded, clients should back off exponentially */
+  final case object Overloaded extends PartyAllocationResult {
```
Comment (stefanobaghino-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PartyAllocationResult.scala]:
```suggestion
  case object Overloaded extends PartyAllocationResult {
```
`final` is redundant on `object`s.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PartyAllocationResult.scala [hunk 4]
```diff
@@ -16,19 +16,30 @@ object PartyAllocationResult {
...
+    override def description: String = "System is overloaded, please try again later"
+  }
+
   /** Synchronous party allocation is not supported */
   final case object NotSupported extends PartyAllocationResult {
```
Comment (stefanobaghino-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PartyAllocationResult.scala]:
```suggestion
  case object NotSupported extends PartyAllocationResult {
```

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PartyAllocationResult.scala [hunk 5]
```diff
@@ -16,19 +16,30 @@ object PartyAllocationResult {
...
     override def description: String = "Party allocation not supported"
```
Comment (stefanobaghino-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/PartyAllocationResult.scala]:
```suggestion
    override val description: String = "Party allocation not supported"
```

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/UploadPackagesResult.scala [hunk 6]
```diff
@@ -14,14 +14,25 @@ object UploadPackagesResult {
     override def description: String = "Packages successfully uploaded"
   }
 
+  /** The system is overloaded, clients should back off exponentially */
+  final case object Overloaded extends UploadPackagesResult {
```
Comment (stefanobaghino-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/UploadPackagesResult.scala]:
```suggestion
  case object Overloaded extends UploadPackagesResult {
```

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/UploadPackagesResult.scala [hunk 7]
```diff
@@ -14,14 +14,25 @@ object UploadPackagesResult {
...
+    override def description: String = "System is overloaded, please try again later"
```
Comment (stefanobaghino-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/UploadPackagesResult.scala]:
```suggestion
    override val description: String = "System is overloaded, please try again later"
```

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/UploadPackagesResult.scala [hunk 8]
```diff
@@ -14,14 +14,25 @@ object UploadPackagesResult {
...
+  }
+
   /** Synchronous package upload is not supported */
   final case object NotSupported extends UploadPackagesResult {
```
Comment (stefanobaghino-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/UploadPackagesResult.scala]:
```suggestion
  case object NotSupported extends UploadPackagesResult {
```

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/UploadPackagesResult.scala [hunk 9]
```diff
@@ -14,14 +14,25 @@ object UploadPackagesResult {
...
     override def description: String = "Packages upload not supported"
```
Comment (stefanobaghino-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/UploadPackagesResult.scala]:
```suggestion
    override val description: String = "Packages upload not supported"
```

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPackageManagementService.scala [hunk 10]
```diff
@@ -76,6 +76,10 @@ class ApiPackageManagementService(
           .flatMap {
             case UploadPackagesResult.Ok =>
               Future.successful(UploadDarFileResponse())
+            case r @ UploadPackagesResult.Overloaded =>
```
Comment (stefanobaghino-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPackageManagementService.scala]:
```suggestion
            case UploadPackagesResult.Overloaded =>
```
`r` seems to be unused.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPartyManagementService.scala [hunk 11]
```diff
@@ -90,8 +90,12 @@ class ApiPartyManagementService private (
       .flatMap {
         case PartyAllocationResult.Ok(details) =>
           Future.successful(AllocatePartyResponse(Some(mapPartyDetails(details))))
+        case r @ PartyAllocationResult.Overloaded =>
```
Comment (stefanobaghino-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPartyManagementService.scala]:
```suggestion
        case PartyAllocationResult.Overloaded =>
```
Looks unused as well.

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/validation/ErrorFactories.scala [hunk 12]
```diff
@@ -40,6 +40,9 @@ trait ErrorFactories {
   def permissionDenied(description: String): StatusRuntimeException =
```
Comment (miklos-da) [inline ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/validation/ErrorFactories.scala]: This `trait` has a misleading name, i.e., this is only 1 factory (which creates different types of errors) hence it should be named `ErrorFactory` instead.

**Pull request merged**